File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    List<List<Integer>> ans = new ArrayList();
    public List<List<Integer>> permuteUnique(int[] nums) {
        if (nums == null || nums.length == 0) {
            return ans;
        }

        Arrays.sort(nums);
        helper(nums, new ArrayList(), new HashSet());

        return ans;
    }

    private void helper(int[] nums, List<Integer> list, Set<Integer> set) {
        if (list.size() == nums.length) {
            ans.add(new ArrayList(list));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && set.contains(i - 1) && nums[i] == nums[i - 1]) {
                continue;
            }

            if (!set.contains(i)) {
                list.add(nums[i]);
                set.add(i);
                helper(nums, list, set);
                set.remove(i);
                list.remove(list.size() - 1);
            }
        }
    }
}

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def __init__(self):
        self.ans = []

    def permuteUnique(self, nums):
        if not nums:
            return self.ans

        nums.sort()
        self.helper(nums, [], set())
        return self.ans

    def helper(self, nums, path, used):
        if len(path) == len(nums):
            self.ans.append(path[:])
            return

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1] and (i - 1) in used:
                continue

            if i not in used:
                path.append(nums[i])
                used.add(i)
                self.helper(nums, path, used)
                used.remove(i)
                path.pop()